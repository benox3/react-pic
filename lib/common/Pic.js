import React, { Component } from 'react';
import getResponsiveImage from '../utils/getResponsiveImage';
import debounce from '../utils/debounce';
import isElementInView from '../utils/isElementInView';
import ImageBase from '../server/index';

/**
 * Pic Component
 */
export default class Pic extends Component {
  constructor(props) {
    super(props);
    this.state = {
      image: props.images[props.defaultIndex],
      noScriptImage: props.noscriptIndex ?
        props.images[props.noScriptIndex] :
        props.images[props.images.length-1],
      isBlurred: props.shouldBlur
    };

    this.setResponsiveImage = this.setResponsiveImage.bind(this);
    this.inViewHandler = this.inViewHandler.bind(this);
    this.loadHandler = this.loadHandler.bind(this);

    // calls inViewHandler with a debounce
    this.debouncedInViewHandler = debounce(this.inViewHandler.bind(this), 150);

  }
  componentDidMount() {
    this.loadHandler();

    // set responsive image on scroll
    window.addEventListener('scroll', this.debouncedInViewHandler);

    // set responsive image on resize
    window.addEventListener('resize', this.debouncedInViewHandler);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.debouncedInViewHandler);
    window.removeEventListener('resize', this.debouncedInViewHandler);
  }

  loadHandler() {
    // set optimal image if in view or if elected to render out of view on mount
    this.props.renderOutOfView ? this.setResponsiveImage() : this.inViewHandler();
  }

  inViewHandler() {
    if (isElementInView(this.refs.base)) {
      this.setResponsiveImage();
    }
  }

  /**
   * Set the optimal image src
   */
  setResponsiveImage() {
    try {
      const parent = this.refs.base.parentNode;
      const imageSlotWidth = parent.getBoundingClientRect().width;

      const responsiveImage = getResponsiveImage(
        imageSlotWidth,
        this.props.images,
        this.state.image
      );
      this.setState({
        image: responsiveImage,
        isBlurred: false
      });
    } catch (e) {
      // failed to update image
    }
  }

  render() {
    if (!this.state.image) {
      return null;
    }

    return (
      <div ref='base' style={this.props.baseStyle} onLoad={this.loadHandler}>
        <ImageBase {...this.props} {...this.state} />
      </div>
    );
  }
}

/** Default Props */
Pic.defaultProps = {
  alt: '',
  defaultIndex: 0,
  shouldBlur: false,
  blurAmmount: '20px',
  baseStyle: {
    position: 'relative',
    overflow: 'hidden'
  },
  imgStyle: {
    margin: '0 auto',
    maxWidth: '100%',
    width: '100%'
  },
  images: [],
  renderOutOfView: false
};

/** Prop Types */
Pic.propTypes = {
  // The collection of images
  images: function(props, propName, componentName) {
    const collectionLength = props[propName].length;
    if (collectionLength <= 0) {
      return new Error(
        'No images are found in `' + propName + '` supplied to ' +
        componentName + '. Validation failed.'
      );
    }
    for (let i = 0; i < collectionLength; i++) {
      const imageObj = props[propName][i];
      if (typeof imageObj.url !== 'string') {
        return new Error(
          'Invalid or no url found in `' + propName + '` supplied to ' +
          componentName + '. Validation failed.'
        );
      }
      if (typeof imageObj.width !== 'number') {
        return new Error(
          'Invalid or no width found in`' + propName + '` supplied to ' +
          componentName + '. Validation failed.'
        );
      }
    }
  },
  defaultIndex: React.PropTypes.number, // The default image to render
  noscriptIndex: React.PropTypes.number, // The default image to render on noscript
  alt: React.PropTypes.string,
  imgStyle: React.PropTypes.object,
  baseStyle: React.PropTypes.object,
  shouldBlur: React.PropTypes.bool,
  blurAmmount: React.PropTypes.string,
  renderOutOfView: React.PropTypes.bool
};
