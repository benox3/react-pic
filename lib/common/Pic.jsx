import React, { Component, PropTypes } from 'react';
import getResponsiveImage from '../utils/getResponsiveImage';
import debounce from '../utils/debounce';
import isElementInView from '../utils/isElementInView';
import ImageWrapper from '../server/index';

/**
 * Pic Component
 */
export default class Pic extends Component {
  /** Prop Types */
  static propTypes = {
    // The collection of images
    images: PropTypes.arrayOf(
      PropTypes.shape({
        url: PropTypes.string.isRequired,
        width: PropTypes.number.isRequired,
      }).isRequired,
    ).isRequired,
    defaultIndex: PropTypes.number, // The default image to render
    noscriptIndex: PropTypes.number, // The default image to render on noscript
    alt: PropTypes.string,
    imgStyle: PropTypes.objectOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
      ]),
    ),
    baseStyle: PropTypes.objectOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
      ]),
    ),
    shouldBlur: PropTypes.bool,
    blurAmount: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    renderOutOfView: PropTypes.bool,
  };

  /** Default Props */
  static defaultProps = {
    alt: '',
    defaultIndex: 0,
    shouldBlur: false,
    blurAmount: '20px',
    baseStyle: {
      position: 'relative',
      overflow: 'hidden',
    },
    imgStyle: {
      margin: '0 auto',
      maxWidth: '100%',
      width: '100%',
    },
    images: [],
    noscriptIndex: 0,
    renderOutOfView: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      image: props.images[props.defaultIndex],
      noScriptImage: props.noscriptIndex ?
        props.images[props.noScriptIndex] :
        props.images[props.images.length-1],
      isBlurred: props.shouldBlur
    };

    this.setResponsiveImage = this.setResponsiveImage.bind(this);
    this.inViewHandler = this.inViewHandler.bind(this);

    // calls inViewHandler with a debounce
    this.debouncedInViewHandler = debounce(this.inViewHandler.bind(this), 150);
  }

  componentDidMount() {
    this.inViewHandler();

    // set responsive image on scroll
    window.addEventListener('scroll', this.debouncedInViewHandler);

    // set responsive image on resize
    window.addEventListener('resize', this.debouncedInViewHandler);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.debouncedInViewHandler);
    window.removeEventListener('resize', this.debouncedInViewHandler);
  }

  /**
   * Sets responsive image if the element is in view
   */
  inViewHandler() {
    if (this.props.renderOutOfView || isElementInView(this.refs.base)) {
      this.setResponsiveImage();
    }
  }

  /**
   * Set the optimal image src
   */
  setResponsiveImage() {
    try {
      const parent = this.refs.base.parentNode;
      const imageSlotWidth = parent.getBoundingClientRect().width;

      const responsiveImage = getResponsiveImage(
        imageSlotWidth,
        this.props.images,
        this.state.image,
      );
      this.setState({
        image: responsiveImage,
        isBlurred: false,
      });
    } catch (e) {
      // failed to update image
    }
  }

  render() {
    if (!this.state.image) {
      return null;
    }

    return (
      <div ref="base" style={this.props.baseStyle} onLoad={this.inViewHandler}>
        <ImageWrapper {...this.props} {...this.state} />
      </div>
    );
  }
}
