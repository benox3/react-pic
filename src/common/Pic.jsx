import React, { Component, PropTypes } from 'react';
import throttle from 'lodash/throttle';
import debounce from 'lodash/debounce';

import getResponsiveImage from '../utils/getResponsiveImage';
import isElementInView from '../utils/isElementInView';
import ImageWrapper from '../server/index';

/**
 * Pic Component
 */
export default class Pic extends Component {
  /** Prop Types */
  static propTypes = {
    // The collection of images
    images: PropTypes.arrayOf(
      PropTypes.shape({
        url: PropTypes.string.isRequired,
        width: PropTypes.number.isRequired,
      }).isRequired,
    ).isRequired,
    defaultIndex: PropTypes.number, // The default image to render
    noscriptIndex: PropTypes.number, // The default image to render on noscript
    baseStyle: PropTypes.objectOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
      ]),
    ),
    shouldBlur: PropTypes.bool,
    renderOutOfView: PropTypes.bool,
  };

  /** Default Props */
  static defaultProps = {
    alt: '',
    defaultIndex: 0,
    shouldBlur: false,
    blurAmount: '20px',
    baseStyle: {
      position: 'relative',
      overflow: 'hidden',
    },
    imgStyle: {
      margin: '0 auto',
      maxWidth: '100%',
      width: '100%',
    },
    images: [],
    noscriptIndex: 0,
    renderOutOfView: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      image: props.images[props.defaultIndex],
      noscriptImage: props.noscriptIndex ?
        props.images[props.noscriptIndex] :
        props.images[props.images.length - 1],
      isBlurred: props.shouldBlur,
    };
    this.setPicRef = this.setPicRef.bind(this);
    this.setResponsiveImage = this.setResponsiveImage.bind(this);
    this.inViewHandler = this.inViewHandler.bind(this);

    // inViewHandlers
    this.inViewHandler = this.inViewHandler.bind(this);
    this.debouncedInViewHandler = debounce(this.inViewHandler, 150);
    this.throttledInViewHandler = throttle(this.inViewHandler, 150);
  }

  componentDidMount() {
    this.inViewHandler();

    // set responsive image on scroll
    window.addEventListener('scroll', this.throttledInViewHandler);

    // set responsive image on resize
    window.addEventListener('resize', this.debouncedInViewHandler);
  }

  componentWillUnmount() {
    this.throttledInViewHandler.cancel();
    this.debouncedInViewHandler.cancel();
    window.removeEventListener('scroll', this.throttledInViewHandler);
    window.removeEventListener('resize', this.debouncedInViewHandler);
  }

  setPicRef(node) {
    this.picRef = node;
  }

  /**
   * Set the optimal image src
   */
  setResponsiveImage() {
    try {
      const parent = this.picRef.parentNode;
      const imageSlotWidth = parent.getBoundingClientRect().width;

      const responsiveImage = getResponsiveImage(
        imageSlotWidth,
        this.props.images,
        this.state.image,
      );
      this.setState({
        image: responsiveImage,
        isBlurred: false,
      });
    } catch (e) {
      // failed to update image
    }
  }

  /**
   * Sets responsive image if the element is in view
   */
  inViewHandler() {
    if (this.props.renderOutOfView || isElementInView(this.picRef)) {
      this.setResponsiveImage();
    }
  }

  render() {
    if (!this.state.image) {
      return null;
    }

    return (
      <div ref={this.setPicRef} style={this.props.baseStyle} onLoad={this.inViewHandler}>
        <ImageWrapper {...this.props} {...this.state} />
      </div>
    );
  }
}
