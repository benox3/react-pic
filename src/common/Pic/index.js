export { default as Pic } from './Pic';