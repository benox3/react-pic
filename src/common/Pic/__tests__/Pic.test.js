import React from 'react';

import { shallow, mount, render } from 'enzyme';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import snapshotDiff from 'snapshot-diff';

import Pic from '../Pic';

describe('Pic', () => {
  let sandbox;
  let stub;

  beforeEach(() => {
    sandbox = sinon.sandbox.create();
  });

  afterEach(() => {
    sandbox.restore();
    stub && stub.reset();
  });

  describe('rendering', () => {
    it('should render the first image in images', () => {
      const props = {
        alt: 'heart',
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };

      const pic = mount(<Pic {...props} />);

      expect(pic.containsMatchingElement(
        <img
          alt={props.alt}
          src={props.images[0].url}
        />,
      )).to.equal(true);
    });

    it('should render the second image in images', () => {
      const props = {
        alt: 'heart',
        defaultIndex: 1,
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };

      const pic = mount(
        <Pic {...props} />,
      );

      expect(pic.containsMatchingElement(
        <img
          alt={props.alt}
          src={props.images[1].url}
        />,
      )).to.equal(true);
    });

    it('should render the last image in noscript', () => {
      const props = {
        alt: 'heart',
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };

      const pic = render(<Pic {...props} />);

      expect(pic.find('noscript img').length).to.equal(1);
      expect(pic.find('noscript img').attr('src')).to.equal(props.images[1].url);
    });

    it('should not render image if props are not passed', () => {
      const pic = shallow(<Pic />);

      expect(pic.find('img')).to.have.length(0);
    });

    describe('conditional rendering', () => {
      it('should not render if a url in images is not a string', () => {
        stub = sinon.stub(console, 'error');
        const props = {
          alt: 'heart',
          images: [
            {
              width: 290,
              url: 'http://placehold.it/290?text=♥',
            },
            {
              width: 320,
              url: {},
            },
          ],
        };

        const wrapper = shallow(<Pic {...props} />);
        expect(wrapper.find('img')).to.have.length(0);
      });

      it('should not render if a width in images is not a number', () => {
        const props = {
          alt: 'heart',
          images: [
            {
              width: 290,
              url: 'http://placehold.it/290?text=♥',
            },
            {
              width: 'test',
              url: 'http://placehold.it/290?text=♥',
            },
          ],
        };

        const wrapper = shallow(<Pic {...props} />);
        expect(wrapper.find('img')).to.have.length(0);
      });

      it('should not render if no images passed through', () => {
        const props = {
          alt: 'heart',
          images: [],
        };

        const wrapper = shallow(<Pic {...props} />);
        expect(wrapper.find('img')).to.have.length(0);
      });
    });
  });

  describe('lifecycles', () => {
    it('should check if image is in view once mounted', () => {
      const props = {
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };

      sandbox.spy(Pic.prototype, 'inViewHandler');
      mount(<Pic {...props} />);

      expect(Pic.prototype.inViewHandler.callCount).to.equal(1);
    });

    it('should call componentWillUnmount when unmounted', () => {
      const props = {
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };

      sandbox.spy(Pic.prototype, 'componentWillUnmount');
      const wrapper = mount(<Pic {...props} />);
      wrapper.unmount();
      expect(Pic.prototype.componentWillUnmount.callCount).to.equal(1);
    });

    it('should set optimal image if renderOutOfView is true', () => {
      const props = {
        renderOutOfView: true,
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };

      sandbox.spy(Pic.prototype, 'setResponsiveImage');
      mount(<Pic {...props} />);

      expect(Pic.prototype.setResponsiveImage.callCount).to.equal(1);
    });
  });

  describe('prop variation', () => {
    it('should set blur to false by default', () => {
      const props = {
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };
      const pic = render(<Pic {...props} />);

      expect(pic.find('img').last().attr('style')).to.not.contain('blur');
    });

    it('should set blur style based if shouldBlur prop is true', () => {
      const props = {
        shouldBlur: true,
        blurAmount: '10px',
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };
      const pic = render(<Pic {...props} />);

      expect(pic.find('img').last().attr('style')).to.contain(
        `blur(${props.blurAmount});`,
      );
    });

    it('should set blur style based on blurAmount prop', () => {
      const props = {
        shouldBlur: true,
        blurAmount: '20px',
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };
      const pic = render(<Pic {...props} />);

      expect(
        pic.find('img').last().attr('style'),
      ).to.contain(`blur(${props.blurAmount});`);
    });

    it('should override img style if imgStyle prop is set', () => {
      const props = {
        imgStyle: {
          backgroundColor: 'red',
        },
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };
      const pic = render(<Pic {...props} />);

      expect(pic.find('img').last().attr('style')).to.contain(
        `background-color:${props.imgStyle.backgroundColor};`,
      );
    });

    it('should override container style if baseStyle prop is set', () => {
      const props = {
        baseStyle: {
          backgroundColor: 'red',
        },
        images: [
          {
            width: 290,
            url: 'http://placehold.it/290?text=♥',
          },
          {
            width: 320,
            url: 'http://placehold.it/320?text=♥',
          },
        ],
      };
      const pic = render(<Pic {...props} />);

      expect(pic.find('div').first().attr('style')).to.contain(
        `background-color:${props.baseStyle.backgroundColor};`,
      );
    });
  });

  describe('snapshots', () => {
    const createNodeMock = element => {
      return {
        getBoundingClientRect() {
          return {
            width: 290
          }
        }
      };
    }
    const options = {createNodeMock};

    const props = {
      alt: 'heart',
      images: [
        {
          width: 290,
          url: 'http://placehold.it/290?text=♥',
        },
        {
          width: 320,
          url: 'http://placehold.it/320?text=♥',
        },
      ],
    };

    it('should match base setup', () => {
      expect(
        renderer.create(<Pic {...props} />, options)
      ).toMatchSnapshot();
    });

    it('should match snapshot diff with defaultIndex provided', () => {
      const defaultRender = renderer.create(
        <Pic {...props} />,
        options
      ).toJSON();
      const defaultIndexRender = renderer.create(
        <Pic defaultIndex={1} {...props} />,
        options
      ).toJSON();

      expect(snapshotDiff(
        defaultRender,
        defaultIndexRender,
      )).toMatchSnapshot();
    });
  });
});
