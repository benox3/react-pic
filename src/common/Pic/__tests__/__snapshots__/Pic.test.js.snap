// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pic snapshots should match base setup 1`] = `
<div
  onLoad={[Function]}
  style={
    Object {
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <div>
    <noscript
      dangerouslySetInnerHTML={
        Object {
          "__html": "<img alt='heart' style='position:absolute;z-index:10;margin:0 auto;max-width:100%;width:100%;' src='http://placehold.it/320?text=♥' />",
        }
      }
    />
    <img
      alt="heart"
      src="http://placehold.it/290?text=♥"
      style={
        Object {
          "margin": "0 auto",
          "maxWidth": "100%",
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;

exports[`Pic snapshots should match snapshot diff with defaultIndex provided 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -15,11 +15,11 @@
          }
        }
      />
      <img
        alt=\\"heart\\"
-       src=\\"http://placehold.it/290?text=♥\\"
+       src=\\"http://placehold.it/320?text=♥\\"
        style={
          Object {
            \\"margin\\": \\"0 auto\\",
            \\"maxWidth\\": \\"100%\\",
            \\"width\\": \\"100%\\","
`;
